src/main/java/com/sforce/ws/transport/JdkHttpStreamingTransport.java

```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are permitted provided
+ * that the following conditions are met:
+ *
+ *    Redistributions of source code must retain the above copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ *    Redistributions in binary form must reproduce the above copyright notice, this list of conditions and
+ *    the following disclaimer in the documentation and/or other materials provided with the distribution.
+ *
+ *    Neither the name of salesforce.com, inc. nor the names of its contributors may be used to endorse or
+ *    promote products derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.sforce.ws.transport;
+
+import java.io.*;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.*;
+import java.util.Map.Entry;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+import com.sforce.ws.*;
+import com.sforce.ws.tools.VersionInfo;
+import com.sforce.ws.util.Base64;
+import com.sforce.ws.util.FileUtil;
+
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLContext;
+
+/**
+ * This class is an implementation of Transport using the build in
+ * JDK URLConnection.
+ *
+ * @author http://cheenath.com
+ * @version 1.0
+ * @since 1.0  Nov 30, 2005
+ */
+public class JdkHttpStreamingTransport implements StreamingTransportInterface {
```

Comment:
The JavaDoc for this class says it was written in 2005. Does that mean it is being copied from somewhere else? If so, from where?

Reply:
JdkHttpStreamingTransport.java is largely a copy/paste of JdkHttpTransport.java with the following diffs:

```
$diff JdkHttpTransport.java JdkHttpStreamingTransport.java | more
...
...
73c76
<         return connectLocal(uri, httpHeaders, true);
---
>         return connectLocal(uri, httpHeaders, true,  StreamingTransportInterface.SupportedHttpMethodType.POST, null, null);
79c82
<         return connectLocal(uri, httpHeaders, enableCompression);
---
>         return connectLocal(uri, httpHeaders, enableCompression,  StreamingTransportInterface.SupportedHttpMethodType.POST, null, null);
94c97
<         return connectLocal(uri, header);
---
>         return connectLocal(uri, header, StreamingTransportInterface.SupportedHttpMethodType.POST, null, null)
97,98c100,116
<     private OutputStream connectLocal(String uri, HashMap<String, String> httpHeaders) throws IOException {
<         return connectLocal(uri, httpHeaders, true);
---
>       @Override
>       public OutputStream connect(String uri, HashMap<String, String> httpHeaders, boolean enableCompression,
>                       SupportedHttpMethodType httpMethod) throws IOException {
>         return connectLocal(uri, httpHeaders, enableCompression, httpMethod, null, null);
>       }
> 
>       @Override
>       public void connectStream(String uri, HashMap<String, String> httpHeaders, boolean enableCompression,
>                       SupportedHttpMethodType httpMethod, InputStream contentInputStream, String contentEncoding)
>                       throws IOException {
>         connectLocal(uri, httpHeaders, enableCompression, httpMethod, contentInputStream, contentEncoding);           
>       }
>       
>     private OutputStream connectLocal(String uri, HashMap<String, String> httpHeaders,
>               StreamingTransportInterface.SupportedHttpMethodType httpMethod, InputStream contentInputStream, String contentEncoding)
>             throws IOException {
>         return connectLocal(uri, httpHeaders, true, httpMethod, contentInputStream, contentEncoding);
101c119,120
<     private OutputStream connectLocal(String uri, HashMap<String, String> httpHeaders, boolean enableCompression)
---
>     private OutputStream connectLocal(String uri, HashMap<String, String> httpHeaders, boolean enableCompression,
>               StreamingTransportInterface.SupportedHttpMethodType httpMethod, InputStream contentInputStream, String contentEncoding)
103c122,127
<         return wrapOutput(connectRaw(uri, httpHeaders, enableCompression), enableCompression);
---
>       OutputStream requestOutputStream = connectRaw(uri, httpHeaders, enableCompression, httpMethod, contentInputStream, contentEncoding);
>       if (contentInputStream != null) {
>               // skip output wrapping
>               return null;
>       }
>         return wrapOutput(requestOutputStream, enableCompression);
127c151,152
<     private OutputStream connectRaw(String uri, HashMap<String, String> httpHeaders, boolean enableCompression)
---
>     private OutputStream connectRaw(String uri, HashMap<String, String> httpHeaders, boolean enableCompression,
>               StreamingTransportInterface.SupportedHttpMethodType httpMethod, InputStream contentInputStream, String contentStreamContentEncoding)
130a156,167
>         if (httpMethod == null) {
>               httpMethod = StreamingTransportInterface.SupportedHttpMethodType.POST;
>         }
>         
>         if (contentInputStream != null) {
>               enableCompression = false;
>             connection.addRequestProperty("Content-Encoding", contentStreamContentEncoding);
>             this.contentInputStream = new BufferedInputStream(contentInputStream);
>         } else {
>               this.contentInputStream = null;
>         }
>         
136c173
<         connection.setRequestMethod("POST");
---
>         connection.setRequestMethod(httpMethod.toString());
207a245,254
>       if (this.contentInputStream != null) {
>               // send the request content
>             byte[] buffer = new byte[2048];
>             BufferedOutputStream bos = new BufferedOutputStream(this.connection.getOutputStream());
>               for(int len; (len = this.contentInputStream.read(buffer)) > 0;) {
>                   bos.write(buffer, 0, len);
>               }
>               bos.close();
>       }

```